src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class ReadOnlyTagHelperAttributes<TAttributeType> : IReadOnlyList<TAttributeType>
```

Comment:
@Eilon interesting idea

we could go with various primary views of the attribute collection:
1. a list of `TagHelperAttribute` instances where each `TagHelperAttribute` has a `Name`
2. a list of `KeyValuePair<string, TagHelperAttribute>` instances where `TagHelperAttribute` contains information about the value (e.g. whether value is missing, meaning the attribute was minimized) but no `Name`
3. some partial implementation of the `IList<T>` interface

in addition, we need a few convenience methods for by-name additions, lookups, et cetera.  this secondary view may be:
1. an `ILookup<string, TagHelperAttribute>`
2. an `IDictionary<string, TagHelperAttribute>` (hmm, this together with 2. above might avoid some of the covariance issues @NTaylorMullen hit)
3. some partial implementation of either interface

I'd been thinking the convenience methods felt like an `IDictionary<string, TagHelperAttribute>` facade because they explicitly ignore the 1:many nature of the underlying collection.  an `ILookup<string, TagHelperAttribute>` is closer to the truth but not that helpful for developers who want to ignore duplicate attributes.

thoughts?